File: app/src/main/java/com/commonpepper/photosen/ui/activities/CropActivity.java
package com.commonpepper.photosen.ui.activities;

import android.app.WallpaperManager;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.commonpepper.photosen.R;
import com.theartofdev.edmodo.cropper.CropImageView;

import java.io.IOException;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

public class CropActivity extends AppCompatActivity {
    public static final String TAG_URISTR = "tag_uri_str";
    private CropImageView cropView;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_crop);

        Toolbar toolbar = findViewById(R.id.toolbar);
        cropView = findViewById(R.id.cropView);

        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_close_white_24dp);
        getSupportActionBar().setTitle(getString(R.string.crop_wallpaper));

        int width = WallpaperManager.getInstance(this).getDesiredMinimumWidth();
        int height = WallpaperManager.getInstance(this).getDesiredMinimumHeight();
//        String path = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
//                .getPath() + File.separator + "wallpaper.jpg";
//        wallpaperUri = FileProvider.getUriForFile(this, Photosen.PACKAGE_NAME + ".fileprovider", new File(path));

        String uriStr = getIntent().getStringExtra(TAG_URISTR);
        Uri sourceUri = Uri.parse(uriStr);
        cropView.setImageUriAsync(sourceUri);
        cropView.setAspectRatio(width, height);

        cropView.setOnCropImageCompleteListener((view, result) -> {
            if (result.getError() == null) {
                try {
                    Bitmap bitmap = result.getBitmap();
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                        WallpaperManager.getInstance(this).setBitmap(bitmap, null, true,
                                WallpaperManager.FLAG_SYSTEM | WallpaperManager.FLAG_LOCK);
                    } else {
                        WallpaperManager.getInstance(this).setBitmap(bitmap);
                    }
                    finish();
                } catch (IOException e) {
                    showError(e);
                }
            } else {
                showError(result.getError());
            }
        });
    }

    private void showError(Exception e) {
        Toast.makeText(CropActivity.this, getString(R.string.error) + ": " + e.toString(),
                Toast.LENGTH_SHORT).show();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.crop_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.crop_done) {
            cropView.getCroppedImageAsync();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return true;
    }
}

File: app/src/main/java/com/commonpepper/photosen/ui/activities/CropActivity.kt
package com.commonpepper.photosen.ui.activities

import android.app.WallpaperManager
import android.graphics.Bitmap
import android.net.Uri
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.commonpepper.photosen.R
import com.commonpepper.photosen.R.*
import com.theartofdev.edmodo.cropper.CropImageView
import com.theartofdev.edmodo.cropper.CropImageView.CropResult
import kotlinx.android.synthetic.main.activity_crop.*
import java.io.IOException

class CropActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout.activity_crop)
        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
            setHomeAsUpIndicator(drawable.ic_close_white_24dp)
            title = getString(string.crop_wallpaper)
        }
        val width = WallpaperManager.getInstance(this).desiredMinimumWidth
        val height = WallpaperManager.getInstance(this).desiredMinimumHeight
        val uriStr: String? = intent.getStringExtra(TAG_URISTR)
        val sourceUri: Uri? = Uri.parse(uriStr)
        cropView.setImageUriAsync(sourceUri)
        cropView.setAspectRatio(width, height)
        cropView.setOnCropImageCompleteListener { _, result: CropResult ->
            if (result.error == null) {
                try {
                    val bitmap: Bitmap? = result.bitmap
                    if (VERSION.SDK_INT >= VERSION_CODES.N) {
                        WallpaperManager.getInstance(this).setBitmap(bitmap, null, true,
                                WallpaperManager.FLAG_SYSTEM or WallpaperManager.FLAG_LOCK)
                    } else {
                        WallpaperManager.getInstance(this).setBitmap(bitmap)
                    }
                    finish()
                } catch (e: IOException) {
                    showError(e)
                }
            } else {
                showError(result.error)
            }
        }
    }

    private fun showError(e: Exception) {
        Toast.makeText(this@CropActivity, getString(string.error) + ": " + e.toString(),
                Toast.LENGTH_SHORT).show()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.crop_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.crop_done) {
            cropView!!.getCroppedImageAsync()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }

    companion object {
        const val TAG_URISTR = "tag_uri_str"
    }
}
